java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java

```diff
@@ -357,379 +349,87 @@ public static void jdbcToArrowVectors(ResultSet rs, VectorSchemaRoot root, Calen
   public static void jdbcToArrowVectors(ResultSet rs, VectorSchemaRoot root, JdbcToArrowConfig config)
       throws SQLException, IOException {
 
-    Preconditions.checkNotNull(rs, "JDBC ResultSet object can't be null");
-    Preconditions.checkNotNull(root, "VectorSchemaRoot object can't be null");
-    Preconditions.checkNotNull(config, "JDBC-to-Arrow configuration cannot be null");
-
     ResultSetMetaData rsmd = rs.getMetaData();
     int columnCount = rsmd.getColumnCount();
-
     allocateVectors(root, DEFAULT_BUFFER_SIZE);
 
+    JdbcConsumer[] consumers = new JdbcConsumer[columnCount];
+    for (int i = 1; i <= columnCount; i++) {
+      consumers[i - 1] = getConsumer(rs, i, rs.getMetaData().getColumnType(i),
+          root.getVector(rsmd.getColumnName(i)), config);
+    }
+
     int rowCount = 0;
     while (rs.next()) {
-      for (int i = 1; i <= columnCount; i++) {
-        jdbcToFieldVector(
-            rs,
-            i,
-            rs.getMetaData().getColumnType(i),
-            rowCount,
-            root.getVector(rsmd.getColumnName(i)),
-            config);
+      for (JdbcConsumer consumer : consumers) {
+        consumer.consume(rs);
       }
       rowCount++;
     }
     root.setRowCount(rowCount);
-  }
 
-  private static void jdbcToFieldVector(
-      ResultSet rs,
-      int columnIndex,
-      int jdbcColType,
-      int rowCount,
-      FieldVector vector,
-      JdbcToArrowConfig config)
-          throws SQLException, IOException {
+    // clean up
+    for (JdbcConsumer consumer : consumers) {
```

Comment:
there is a leak here if an exception is thrown?

Reply:
Right.